### PR TITLE
add middleware for artifact with subject

### PR DESCRIPTION
**Pull request by wy65701436**
As for the distribution spec 1.1, it supports client to push an manifest with subject field. By leverging this fidle, harbor could build up the linkage between the subject artifact and it's accessories.

Thank you for contributing to Harbor!

# Comprehensive Summary of your change

# Issue being fixed
Fixes #(issue)

Please indicate you've done the following:
- [x] Well Written Title and Summary of the PR
- [x] Label the PR as needed. "release-note/ignore-for-release, release-note/new-feature, release-note/update, release-note/enhancement, release-note/community, release-note/breaking-change, release-note/docs, release-note/infra, release-note/deprecation"
- [x] Accepted the DCO. Commits without the DCO will delay acceptance.
- [x] Made sure tests are passing and test coverage is added if needed.
- [ ] Considered the docs impact and opened a new docs issue or PR with docs changes if needed in [website repository](https://github.com/goharbor/website).

### CODE REVIEW

File: src/pkg/accessory/model/subject/subject.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+	"github.com/goharbor/harbor/src/pkg/accessory/model"
+	"github.com/goharbor/harbor/src/pkg/accessory/model/base"
+)
+
+// Subject model
+type Subject struct {
+	base.Default
+}
+
+// Kind gives the reference type of nydus accelerator.
```
Comment (chlins) [inline src/pkg/accessory/model/subject/subject.go:27]: `nydus accelerator` -> `subject`

File: src/pkg/accessory/model/subject/subject.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+func (s *Subject) Kind() string {
+	return model.RefHard
+}
+
+// IsHard ...
+func (s *Subject) IsHard() bool {
+	return true
+}
+
+// New returns nydus accelerator
```
Comment (chlins) [inline src/pkg/accessory/model/subject/subject.go:37]: ditto

File: src/server/middleware/subject/subject.go [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+		ctx := r.Context()
+		logger := log.G(ctx).WithFields(log.Fields{"middleware": "subject"})
+
+		none := lib.ArtifactInfo{}
+		info := lib.GetArtifactInfo(ctx)
+		if info == none {
+			return errors.New("artifactinfo middleware required before this middleware").WithCode(errors.NotFoundCode)
+		}
+
+		body, err := io.ReadAll(r.Body)
```
Comment (chlins) [inline src/server/middleware/subject/subject.go:73]: Do we need to close the body after read in this place?

**Review by chlins — APPROVED** [review]
lgtm

File: src/server/middleware/subject/subject.go (hunk 3 above)
Comment (wy65701436) [reply src/server/middleware/subject/subject.go:73]: for put manifest request, it is a no close request which was set by put_mainfest_middlewale

File: src/server/middleware/subject/subject.go [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+				logger.Errorf("failed to get subject artifact: %s, error: %v", mf.Subject.Digest, err)
+				return err
+			}
+			art, err := artifact.Ctl.GetByReference(ctx, info.Repository, info.Reference, nil)
+			if err != nil {
+				logger.Errorf("failed to get artifact with subject field: %s, error: %v", info.Reference, err)
+				return err
+			}
+
+			if err := orm.WithTransaction(func(ctx context.Context) error {
```
Comment (stonezdj) [inline src/server/middleware/subject/subject.go:96]: Because it is already in the scope of TransactionMiddleware, should we create a new transaction?

Comment (wy65701436) [reply src/server/middleware/subject/subject.go:96]: it is to avoid the errors.IsConflictErr().

**Review by stonezdj — APPROVED** [review]

**Pull request merged**